src/ray/rpc/raylet/raylet_client.h

```diff
@@ -196,8 +195,33 @@ class RayletClient {
   /// for this worker. Each pair consists of the resource ID and the fraction
   /// of that resource allocated for this worker.
   ResourceMappingType resource_ids_;
-  /// The connection to the raylet server.
-  std::unique_ptr<RayletConnection> conn_;
+
+  /// The gRPC-generated stub.
+  std::unique_ptr<RayletService::Stub> stub_;
+
+  /// Service for handling reply.
+  boost::asio::io_service main_service_;
+
+  /// Asio work for main service.
+  boost::asio::io_service::work work_;
+
+  /// The `ClientCallManager` used for managing requests.
+  ClientCallManager client_call_manager_;
+
+  /// The thread used to handle reply.
+  std::thread rpc_thread_;
+
+  /// Heartbeat timer.
+  boost::asio::deadline_timer heartbeat_timer_;
+
+  /// Mutex for to protect resource_ids_.
+  mutable std::mutex resource_ids_lock_;
```

Comment:
### According to the previous implementation I think the raylet client is a thread-safe client. Assuming the resource_ is using in a thread, and we are getting task in another thread and calling the clear() method, it might crash.

Reply:
Yes, I have removed it.